src/librustc_trans/trans/mir/block.rs

```diff
@@ -41,8 +53,21 @@ impl<'bcx, 'tcx> MirContext<'bcx, 'tcx> {
         debug!("trans_block: terminator: {:?}", data.terminator());
 
         match *data.terminator() {
+            mir::Terminator::Resume => {
+                if let Some(cleanup_pad) = cleanup_pad {
+                    bcx.cleanup_ret(cleanup_pad, None);
+                } else {
+                    let ps = self.get_personality_slot(&bcx);
+                    let lp = bcx.load(ps);
+                    bcx.with_block(|bcx| {
+                        base::call_lifetime_end(bcx, ps);
+                        base::trans_unwind_resume(bcx, lp);
+                    });
+                }
+            }
+
             mir::Terminator::Goto { target } => {
-                bcx.br(self.llblock(target));
+                funclet_br(bcx, self.llblock(target));
```

Comment:
On Thu, Feb 18, 2016 at 01:42:31AM -0800, Simonas Kazlauskas wrote:
> I feel like these restrictions could be enforced by a pass similar to
> typeck. Actually we could just rename typeck to sanityck and throw various
> validity checks into there.

I would prefer to keep *type checking* distinct from *sanity
checking*, but yes. (In particular, we will need to do actual
type-checking when we move to non-lexical lifetimes.)